### ISSUE TITLE
Autocommand event when updates finish

### ISSUE THREAD

**Issue by tralph3** [nvim-treesitter/nvim-treesitter#2900]
**Is your feature request related to a problem? Please describe.**
I want to be able to update neovim and all its plugins automatically. For that, I would need to be able to detect when the updates finish to be able to exit.

**Describe the solution you'd like**
If this plugin would emit an event for an autocommand to grab, then I could detect when it finishes updating and I could exit neovim automatically.

**Describe alternatives you've considered**
Packer already does this. It emits the "User PackerComplete" event. In order to make that work with treesitter, I have enabled synchrnous updating, which makes the update process slow, but makes Packer wait for treesitter to finish updating before it exits, so I can use Packer's event to quit neovim. Ideally, I would leave the asynchronous feature on, and exit when both Packer and treesitter have emitted their events.

**Comment by fsouza** [comment]
You can use `TSUpdateSync`. This is how I do it:

```
nvim --headless -E +'TSUpdateSync' +'quit'
```

To be honest, an ideal solution would be something like:

```
nvim --headless -E +'TSUpdateSync|quit' +'cq'
```

But TSUpdateSync doesn't support that.

**Comment by ecogest** [comment]
> You can use `TSUpdateSync`. This is how I do it:
> 
> ```
> nvim --headless -E +'TSUpdateSync' +'quit'
> ```
> 
> To be honest, an ideal solution would be something like:
> 
> ```
> nvim --headless -E +'TSUpdateSync|quit' +'cq'
> ```
> 
> But TSUpdateSync doesn't support that.

Did not work for me, in spite of 'TSUpdateSync', nvim quits before the update is done.

The only temporary fix for me is to add some sleep in between:
`nvim --headless -c 'TSUpdateSync' -c 'sleep 20' -c 'qa'`
This is not elegant though... A user autocommand event like in packer.nvim would indeed be great!

**Comment by fsouza** [comment]
@ecogest I imagine you're using it for installing new parsers, not update? TSUpdateSync definitely works, but it doesn't install new parsers, it only updates existing parsers.

**Comment by ecogest** [comment]
@fsouza Indeed, I was trying to install the parsers! my mistake ;)
Still, I also tried to install the parsers with `TSInstallSync`, and I can't get them all installed without a sleep. (I guess this should work very similarly to `TSUpdateSync`...)

**Comment by fsouza** [comment]
That's a good catch though. Perhaps there's a bug in TSInstall or something like that.

**Comment by shaunsingh** [comment]
Was there any update on this, `TSInstall` and `TSInstallSync` still require a `sleep` to work in headless mode

**Comment by theHamsta** [comment]
No, there shouldn't be any sleep required. Our CI uses TSInstallSync

**Comment by tkf144** [comment]
> Was there any update on this, `TSInstall` and `TSInstallSync` still require a `sleep` to work in headless mode

I landed in this issue trying to figure out how to achieve synchronous install of parsers supplied via `ensure_installed`, for bootstrapping of Neovim, plugins, etc in a devcontainer.

Per the [README.MD](https://github.com/nvim-treesitter/nvim-treesitter/blob/master/README.md?plain=1#L112), see `sync_install = true` which seems to be used exclusively [in configs.lua](https://github.com/nvim-treesitter/nvim-treesitter/blob/master/lua/nvim-treesitter/configs.lua#L425).

I'm now just starting Neovim without any `:TSInstallSync`, i.e. `nvim --headless +quitall`. The parsers are installed, albeit slowly, but far more reliably too; there was a 4 in 5 chance that the async parser downloads would be corrupted and the TS install would fail.

If one is not using `ensure_installed`, or doesn't want to `require("nvim-treesitter").setup()` with `sync_install`, one could try referencing the [section in configs.lua](https://github.com/nvim-treesitter/nvim-treesitter/blob/master/lua/nvim-treesitter/configs.lua#L426) (I've not tried this):

```lua
require("nvim-treesitter.install").ensure_installed_sync({ "c", "lua" })
```
